Run cpu intensive logic in worker threads

e00E opened the pull request:
Previously only one thread was being used but handling indvidual
requests can take seconds on cpu intensive work.

This is waiting for https://github.com/gnosis/dex-contracts/pull/662 .

e00E committed:
Run cpu intensive logic in worker threads

Previously only one thread was being used but handling indvidual
requests can take seconds on cpu intensive work.

e00E commented on package.json:
```diff
@@ -27,13 +27,15 @@
   },
   "dependencies": {
-    "@gnosis.pm/dex-contracts": "^0.2.10",
+    "@gnosis.pm/dex-contracts": "file:/home/vk/Gnosis/dex-contracts",
```
TODO: remove this and point to next released version

e00E commented on src/index.ts:
```diff
@@ -59,42 +61,46 @@ app.use(morgan('tiny'))
 
+const pool = workerpool.pool(path.join(__dirname, '/worker.js'), {
+  minWorkers: "max",
+  maxWorkers: 10
```
TODO: What number of workers? Probably make it command line argument.

fleupold replied:
yes, probably some power of two? Maybe 8 for now?

nlordell replied:
I would default to `NUM_CPUS` and have it be a command line argument.

nlordell replied:
```
const os = require('os')
const numCpus = os.cpus().length
```

fleupold commented on src/worker.ts:
```diff
@@ -0,0 +1,28 @@
+  const sellAmount = new BN(quoteAmount);
+  // The orderbook API only allows us to compute a price for selling base tokens. Here, we want to sell quote tokens.
+  // Thus, we invert the orderbook, sell the requested amount (now in base) and multiply the computed price with the
+  // sell amount in order to compute the buy amount in the original base token. A small rounding buffer is considered
+  // to increase the chances of matching the order if placed.
```
Can we either apply the rounding buffer here (rather than in the handler) or move all price estimation logic back into the handler. The only expensive thing should be computing the transitive orderbbok since this has quadratic runtime.

nlordell replied:
Then the worker pool is only used for computing the transitive orderbook? That sounds OK to me.

fleupold reviewed (approved):
Nice, looks really good. Were you able to do some local tests to see how it impacts concurrent request performance?

My feeling is that computing the orderbook is the only really expensive part, so we could just have one worker function that is reused by both handlers (and the buyAmount is then computed back on the main-thread in O(n)).
Otherwise, can we move the tiny bit of remaining buyAmount computation completely in the worker thread and leave the handlers free of logic?

Adding @anxolin as he was doing some benchmarks himself today (he is using a tool called [siege](https://github.com/JoeDog/siege)) and might have feedback on the approach (given he knows node very well).

nlordell commented on package.json:
```diff
@@ -27,13 +27,15 @@
   "dependencies": {
-    "@gnosis.pm/dex-contracts": "^0.2.10",
+    "@gnosis.pm/dex-contracts": "file:/home/vk/Gnosis/dex-contracts",
+    "@types/workerpool": "^5.0.1",
     "@types/yargs": "^15.0.4",
```
nit: These should be `devDependencies` since they are only used for compiling.

e00E replied:
And it only needs the types during compilation but the real workerpool when running. I see.

nlordell replied:
:+1: The `@types/*` packages typically only contain `.d.ts` files which are type definitions for describing JS libraries, but don't actually contain any code or a module that gets included at runtime.

nlordell commented on src/index.ts:
```diff
@@ -1,12 +1,14 @@
-import { Fraction, transitiveOrderbook } from '@gnosis.pm/dex-contracts'
+import { Fraction} from '@gnosis.pm/dex-contracts'
```
ubernit: formatting is a bit off.

e00E replied:
How do I auto format in javascript? I tried to use vscode but it changed formatting for lines I didn't touch too and it didn't seem to ever split long lines.

nlordell replied:
Maybe @fleupold can give you some pointers, I have no idea

fleupold replied:
The .eslintrc.js should have the formatting definitions. VSCode does split long lines for me (maybe you need to install the eslint package for that to work).

nlordell reviewed (approved):
This is awesome!!! Maybe we can include an e2e test that benchmarks, say, 1000 concurrent requests so we get some unscientific indication on performance.

nlordell commented on src/worker.ts:
```diff
@@ -0,0 +1,28 @@
+  return result;
+}
+
+function estimatedBuyAmount(orderbooks_object: object, base: string, quote: string, hops: number, quoteAmount: string) {
```
Maybe the `OrderbookFetcher` is responsible for pushing the new orderbook to each worker instead of serializing and sending everytime?

I think that is a separate improvement for another PR though.

fleupold replied:
When I briefly tested the workers weren't running at all times and thus able to capture state but rather spawned whenever a request came in. Thus I wasn't able to "store" the orderbook inside the worker.

nlordell replied:
I wonder if this is a `workerpool` thing because I was under the impression workers that had message handlers setup would run until explicitly closed. It's been SO LONG since I've done any worker related work so my memory might be off.

e00E replied:
Iirc when I was looking for libraries some of them used message passing and some had this abstraction of function call (even though internally they still do message passing). Since I picked this library I will stay with it.

e00E committed:
Update with dex-contracts

e00E committed:
make dev dependdenciest

e00E committed:
Specify number of threads

e00E commented on src/utilities.ts:
```diff
@@ -14,7 +14,14 @@ export function getHops(request: Request, maxHops: number) {
 }
 
 export function sortOrderbookBySerializedPrice(orderbook: Orderbook) {
-  const json = orderbook.toJSON()
+  // TODO: previously one offer looked liked this in json
+  // {"price":9090936436582474,"volume":0.002000049787682}
+  // but now it looks like this
+  // {"volume":{"numerator":"e8d620e991","denominator":"01c6bf52634000"},
+  // "price":{"numerator":"05679e9e863369d800","denominator":"2ad7"}}
+  // Is it important that it doesn't change format? The documentation doesn't specify
+  // the format.
+  const json = orderbook.getOffers()
   json.bids.sort(sortDescending)
   json.asks.sort(sortAscending)
   return json
```
This depends on if other services rely on this format. If they do then I should change it to the original format. Do they?

fleupold replied:
Yes, the frontend relies on having volume and price as a number and not an object type. The e2e test should already be testing that. Do they not?

e00E commented on src/worker.ts:
```diff
@@ -0,0 +1,45 @@
+  // to increase the chances of matching the order if placed.
+  const estimatedPrice = transitive.inverted().priceToSellBaseToken(new BN(sellAmount))
+  console.log(estimatedPrice)
+  if (estimatedPrice !== undefined) {
+    const buyAmountInBase =
+      (1 - priceRoundingBuffer) * new Fraction(sellAmount, 1).mul(estimatedPrice).toNumber()
+    console.log(buyAmountInBase)
+    return {
+      baseTokenId: base,
+      quoteTokenId: quote,
+      buyAmountInBase: buyAmountInBase,
+      sellAmountInQuote: sellAmount.toNumber(),
+    }
+  } else {
+    // TODO: in case there is no result (estimedPrice == null) what should we return?
+    // I think previously it finished the request with req.end() without setting anything but now
+    // we set an empty object. What is more correct?
+    return {}
+  }
```
I prefer the empty object.

fleupold replied:
Actually I think it would be correct to set buyAmountInBase to 0. Either the token does not exist or there is not enough liquidity, in both cases you are expected to get 0 of the tokens you want to buy. This would also normalize the return types more.

nlordell commented:
We might need to deal with too many incoming requests. They might just pile up on the worker pool and it might get so backed up that it will stop responding to new requests.

Are there some timeout option we can set there?

e00E committed:
temp

e00E committed:
Merge remote-tracking branch 'origin/master' into threads

e00E committed:
Use 0 if estimatedPrice is undefined

e00E committed:
temp

e00E commented:
> We might need to deal with too many incoming requests. They might just pile up on the worker pool and it might get so backed up that it will stop responding to new requests.

I have set maxQueueSize which errors if the queue is larger.

e00E commented on src/index.ts:
```diff
@@ -59,54 +64,48 @@ app.use(morgan('tiny'))
+const poolOptions = {
+  minWorkers: argv['num-threads'],
+  maxWorkers: argv['num-threads'],
+  maxQueueSize: argv['num-threads'] * 2,
```
todo configurable

e00E committed:
make queue size configurable

e00E commented:
In my test this scaled with close to the number of threads, a bit slower which I attribute to my laptop's cpu.

e00E committed:
ci

e00E commented:
When I run it the usual way it works but when testing I get
```
Error: Cannot find module '/home/vk/Gnosis/dex-price-estimator/src/worker.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
internal/modules/cjs/loader.js:638
    throw err;
    ^
```
That seems to be something with the test system? Anyone have an idea? Also happens on CI but slightly different orders of messages.

fleupold commented on src/index.ts:
```diff
@@ -59,54 +68,48 @@ app.use(morgan('tiny'))
 app.use('/api/v1/', router)
 const web3 = new Web3(argv['ethereum-node-url'] as string)
 
+console.log('Using ', argv['num-threads'], 'threads')
```
Can we add this to the log statement in line 56?

fleupold commented on src/orderbook_fetcher.ts:
```diff
@@ -29,6 +29,22 @@ export class OrderbookFetcher {
     poll()
   }
 
+  serializeOrderbooks(): object {
+    const o: any = {}
+    this.orderbooks.forEach((value, key) => {
+      o[key] = value.toJSON()
+    })
+    return o
+  }
+
+  static deserializeOrderbooks(o: any): Map<string, Orderbook> {
+    const orderbooks = new Map()
+    for (const [key, value] of Object.entries(o)) {
+      orderbooks.set(key, Orderbook.fromJSON(value))
+    }
+    return orderbooks
```
This is not working on my machine. Fractions are serialized to something like:

`{ numerator: { negative: 0, words: [Array], length: 1, red: null }, denominator: { negative: 0, words: [Array], length: 1, red: null } }` which then cannot be deserialized using `new BN(o.denominator, "hex");`. I need to call `JSON.stringify` & `JSON.parse` to make it work.

Do you have an idea why it is working for you?

nlordell replied:
Maybe different node JS versions?

e00E replied:
So it works on travis and for me but not for you. That IS weird. My yarn version is `1.21.1` and my node version is `10.18.1`.

e00E replied:
Maybe `BN` has released a new semver compatible version (even though it really isn't) that changes how BNs are serialized. We use the default serialization and expect it to serialize a BN to a hex string (see Fraction::fromJSON). On your machine the different BN dependency is used than on my machine and travis.

e00E replied:
BN seems fine. Maybe we do have different node versions and on mine it has to stringify things when passing to the workerpool but not on yours. We can probably fix this by always doing stringify and parse.

fleupold replied:
I'm running yarn 1.17.3 and node v11.15.0

e00E replied:
I pushed a new change that I hope fixes it on your node version too. Also changed the test cleanup slightly which silenced a jest warning I got on my machine.

e00E replied:
@fleupold let me know if it works for you now

fleupold reviewed (commented):
I believe it's not working because the path to the worker script is incorrect under test. I printed it and it evaluates to:

> /Users/felixleupold/Gnosis/dex-price-estimator/src/worker.js

on my machine (it should be under build/worker.js). We can probably work around it for now by using `../build/worker.js` as a path instead. This should work for both `yarn start` and `yarn test`

I will push my local adjustments to this branch in a minute (feel free to revert if you don't like them).

fleupold committed:
making things run on my local machine

e00E committed:
Turn orderbooks into string

This is needed for serialization to work on node version 11
Also make test more reliable

anxolin commented on README.md:
````diff
@@ -50,8 +50,10 @@ yarn start -- --ethereum-node-url <your endpoint>
+
 ```
-export ETHEREUM_NODE_URL=<<your endpoint>>
+export ETHEREUM_NODE_URL=<<your endpoint>> POLL_FREQUENCY=10000000000
````
I didn't see where you use this in the code

e00E replied:
This sets the command line argument of the same name.

anxolin commented on src/index.ts:
```diff
@@ -59,54 +69,47 @@ app.use(morgan('tiny'))
+  maxWorkers: argv['num-threads'],
+  maxQueueSize: argv['max-queue-size'],
+}
+export const pool = workerpool.pool(path.join(__dirname, '../build/worker.js'), poolOptions)
```
Why is the worker imported like this instead of using typescript?

e00E replied:
I don't know and I don't know what the typescript way is. This line originally comes from Felix. Since it works I would like to leave it like this unless you know the correct way to do it.

anxolin commented on src/worker.ts:
```diff
@@ -0,0 +1,45 @@
+  if (estimatedPrice === undefined) {
+    estimatedPrice = Fraction.fromNumber(0)
+  }
+  const buyAmountInBase = (1 - priceRoundingBuffer) * new Fraction(sellAmount, 1).mul(estimatedPrice).toNumber()
```
I think this unrelated to this PR, but I think we should not operate using numbers, right? this loses some precision. `BigNumber.js` can help to do this operations and has rounding strategies when you want to convert in a decimal result after operating.

anxolin reviewed (approved):
Great work! Thanks for tagging me, and sorry for the delay in reading this. 
Well I did a super basic load testing before launch using siege app. For something a bit nicer, I've used in the past JMeter, but I think I would need to start almost from scratch again with that

The good thing about jmeter is that you can build a load test, visualize the result, simulate user sessions or unique users, or execute test like, run the order book changing the parameters a bit so you make sure it can't use any caches.

I didn't run this code, but if you see a performance gain, and you find the pool config that works best, I'm all in with it.

Good job
